File: lua/phone-number.lua
local PhoneNumber = {}
PhoneNumber.__index = PhoneNumber

function PhoneNumber:new(number)
    number = number:gsub("[^%d]", "")
    local default = "0000000000"
    if #number == 11 and number:sub(1, 1) == "1" then
        number = number:sub(2)
    end

    if #number ~= 10 then
        number = default
    end

    local t = {
        number = number,
    }
    return setmetatable(t, PhoneNumber)
end

function PhoneNumber:areaCode()
    return self.number:sub(1, 3)
end

function PhoneNumber.__tostring(phone)
    return string.format("(%s) %s-%s", phone:areaCode(), phone.number:sub(4, 6), phone.number:sub(7))
end

return PhoneNumber

File: go/phone_number.go
package phonenumber

import (
	"errors"
	"fmt"
	"unicode"
)

func Number(phoneNumber string) (string, error) {
	s := []rune{}
	for _, c := range phoneNumber {
		if unicode.IsDigit(c) {
			s = append(s, c)
		}
	}

	if s[0] == '1' {
		s = s[1:]
	}

	if len(s) != 10 {
		return "", errors.New("invalid")
	}

	if s[0] < '2' || s[3] < '2' {
		return "", errors.New("invalid")
	}

	return string(s), nil
}

func AreaCode(phoneNumber string) (string, error) {
	phoneNumber, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}

	return string(phoneNumber[:3]), nil
}

func Format(phoneNumber string) (string, error) {
	phoneNumber, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("(%s) %s-%s", string(phoneNumber[:3]), string(phoneNumber[3:6]), string(phoneNumber[6:])), nil
}
